File: gas_station.cpp
#include <vector>
#include <iostream>
using namespace std;
class Solution {
public:
    int canCompleteCircuit(vector<int> &gas, vector<int> &cost) {
       bool found = false; 
       int i = 0, j = 0;
       int sum = 0;
       int len = gas.size();
       int step = 0;
       while (step < len * 2) {
          sum += (gas[i] - cost[i]);
          // cout << gas[i] - cost[i] << endl;
          // cout << sum << endl;
          if (sum < 0) {
            sum = 0;
            i = (i + 1) % len;
            j = i;
            found = false;
          } else {
            i = (i + 1) % len;
            found = true;
          }
          step++;
       }
       // cout << "sum: " << sum << endl; 
       if (sum > 0 || (sum == 0 && found))
         return j;
       else
         return -1;
    }
};
int main() {
  vector<int> gas {2, 4};
  vector<int> cost {3, 4};
  Solution sol;
  cout << sol.canCompleteCircuit(gas, cost) << endl;
}

File: gas_station.py
class Solution:
    # @param gas, a list of integers
    # @param cost, a list of integers
    # @return an integer
    def canCompleteCircuit(self, gas, cost):
        start = 0
        count = 0
        length = 0
        i = start
        s = 0
        if len(gas) == 1:
            if gas[0] < cost[0]:
                return -1
            else:
                return 0
        while count < 2 * len(gas):
            # print "i = ", i
            s += gas[i] - cost[i]
            # print "s = ", s
            if s < 0:
                start = (i + 1) % len(gas)
                length = 0
                # print "moving start to ", (i + 1) % len(gas)
                s = 0
            else:
                length += 1
            i = (i + 1) % len(gas)
            count += 1
            # print "length: ", length
            if length == len(gas):
                return start
        return -1

sol = Solution()
gas = [1, 2, 3, 3]
cost = [2, 1, 5, 1]
print sol.canCompleteCircuit(gas, cost)

